File: src/routes/apiKeys.js
const express = require('express');
const router = express.Router();
const { v4: uuidv4 } = require('uuid');
const { db } = require('../db');
const adminAuth = require('../middleware/adminAuth');

// Create new API key (admin only)
router.post('/', adminAuth, async (req, res) => {
    try {
        const { name, hourlyLimit } = req.body;
        
        if (!name) {
            return res.status(400).json({ error: 'Name is required' });
        }

        const id = uuidv4();
        const apiKey = uuidv4(); // In production, you might want a more secure key generation method
        
        await db.prepare(`
            INSERT INTO api_keys (id, name, api_key, hourly_limit)
            VALUES (?, ?, ?, ?)
        `).run([id, name, apiKey, hourlyLimit || 100]);

        res.status(201).json({ 
            id,
            name,
            apiKey,
            hourlyLimit: hourlyLimit || 100
        });
    } catch (err) {
        console.error('Error creating API key:', err);
        res.status(500).json({ error: 'Failed to create API key' });
    }
});

// List all API keys (admin only)
router.get('/', adminAuth, async (req, res) => {
    try {
        const keys = await db.prepare(`
            SELECT id, name, api_key, hourly_limit, created_at, request_count
            FROM api_keys
            ORDER BY created_at DESC
        `).all();
        
        res.json(keys);
    } catch (err) {
        console.error('Error listing API keys:', err);
        res.status(500).json({ error: 'Failed to list API keys' });
    }
});

// Delete API key (admin only)
router.delete('/:id', adminAuth, async (req, res) => {
    try {
        await db.prepare('DELETE FROM api_keys WHERE id = ?').run([req.params.id]);
        res.json({ message: 'API key deleted successfully' });
    } catch (err) {
        console.error('Error deleting API key:', err);
        res.status(500).json({ error: 'Failed to delete API key' });
    }
});

module.exports = router;

File: rust-src/src/routes/api_keys.rs
use actix_web::{web, HttpResponse};
use uuid::Uuid;

use crate::error::AppError;
use crate::middleware::admin_auth::AdminUser;
use crate::AppState;

#[derive(serde::Deserialize)]
pub struct CreateApiKeyRequest {
    pub name: Option<String>,
    #[serde(rename = "hourlyLimit")]
    pub hourly_limit: Option<i32>,
}

/// POST /api/keys — Create new API key (admin only)
async fn create_api_key(
    state: web::Data<AppState>,
    _admin: AdminUser,
    body: web::Json<CreateApiKeyRequest>,
) -> Result<HttpResponse, AppError> {
    let name = body
        .name
        .as_deref()
        .filter(|n| !n.is_empty())
        .ok_or_else(|| AppError::BadRequest("Name is required".to_string()))?;

    let id = Uuid::new_v4().to_string();
    let api_key = Uuid::new_v4().to_string();
    let hourly_limit = body.hourly_limit.unwrap_or(100);

    sqlx::query(
        "INSERT INTO api_keys (id, name, api_key, hourly_limit) VALUES (?, ?, ?, ?)",
    )
    .bind(&id)
    .bind(name)
    .bind(&api_key)
    .bind(hourly_limit)
    .execute(&state.pool)
    .await?;

    Ok(HttpResponse::Created().json(serde_json::json!({
        "id": id,
        "name": name,
        "apiKey": api_key,
        "hourlyLimit": hourly_limit
    })))
}

/// GET /api/keys — List all API keys (admin only)
async fn list_api_keys(
    state: web::Data<AppState>,
    _admin: AdminUser,
) -> Result<HttpResponse, AppError> {
    let keys: Vec<crate::models::api_key::ApiKey> = sqlx::query_as(
        "SELECT id, name, api_key, hourly_limit, created_at, last_reset, request_count FROM api_keys ORDER BY created_at DESC",
    )
    .fetch_all(&state.pool)
    .await?;

    Ok(HttpResponse::Ok().json(keys))
}

/// DELETE /api/keys/{id} — Delete API key (admin only)
async fn delete_api_key(
    state: web::Data<AppState>,
    _admin: AdminUser,
    path: web::Path<String>,
) -> Result<HttpResponse, AppError> {
    let id = path.into_inner();

    sqlx::query("DELETE FROM api_keys WHERE id = ?")
        .bind(&id)
        .execute(&state.pool)
        .await?;

    Ok(HttpResponse::Ok().json(serde_json::json!({
        "message": "API key deleted successfully"
    })))
}

pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::scope("/keys")
            .route("", web::post().to(create_api_key))
            .route("", web::get().to(list_api_keys))
            .route("/{id}", web::delete().to(delete_api_key)),
    );
}
